tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15169,6 +15169,60 @@ static bool nodeInputIsCompatible(DXIL::NodeIOKind IOType,
   }
 }
 
+// Diagnose input node record to make sure it has exactly one SV_DispatchGrid
+// semantics. Recursivelly walk all fields on the record and all of its base
+// classes/structs
+void DiagnoseMustHaveOneDispatchGridSemantics(Sema &S,
+                                              CXXRecordDecl *InputRecordDecl,
+                                              SourceLocation &DispatchGridLoc,
+                                              bool &Found) {
+  // Iterate over fields of the input record struct
+  for (auto FieldDecl : InputRecordDecl->fields()) {
+    // Check if any of the fields have SV_DispatchGrid annotation
+    for (const hlsl::UnusualAnnotation *it :
+         FieldDecl->getUnusualAnnotations()) {
+      if (it->getKind() == hlsl::UnusualAnnotation::UA_SemanticDecl) {
+        const hlsl::SemanticDecl *sd = cast<hlsl::SemanticDecl>(it);
+        if (sd->SemanticName.equals("SV_DispatchGrid")) {
```

Comment:
There's been an effort to move away from explicit string compares. I'm not aware of there being any location where this semantic string exists and is stored in a variable, but maybe an issue can be filed so that in the future this could instead look something like 
`.equals( <variable that will always hold "SV_DispatchGrid"> )`

Reply:
I am aware of the effort to move away from explicit string compares, but I am not sure if it makes sense for the HLSL semantics. It is not a fixed set of strings, and the comparisons are much less common than in the other cases for resources, node record types and such. In fact, there are only 3 instances of `semanticName.equals` in the code.

@llvm-beanz - what is your take on this?

Reply:
I think the way we normally look at semantics is using `Semantic::DecomposeNameAndIndex` and `Semantic::GetByName`. It would be nice if the `HLSLSemanticAttr` or `SemanticDecl` parsed the string into the enum and stored the enum value.